// @nrwl/workspace/src/tasks-runner/hasher.spec.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const hasher_1 = require("./hasher");
const hasha = require('hasha');
const fs = require('fs');
jest.mock('hasha');
jest.mock('fs');
describe('Hasher', () => {
    let hashes = { 'yarn.lock': 'yarn.lock.hash' };
    beforeEach(() => {
        hasha.mockImplementation(values => values.join('|'));
        hasha.fromFile.mockImplementation(path => Promise.resolve(hashes[path]));
        fs.statSync.mockReturnValue({ size: 100 });
        fs.readFileSync.mockImplementation(() => JSON.stringify({ dependencies: {}, devDependencies: {} }));
    });
    it('should create project hash', (done) => __awaiter(void 0, void 0, void 0, function* () {
        hashes['/file'] = 'file.hash';
        const hasher = new hasher_1.Hasher({
            nodes: {
                proj: {
                    name: 'proj',
                    type: 'lib',
                    data: { files: [{ file: '/file', ext: '.ts', mtime: 1 }] }
                }
            },
            dependencies: {
                proj: []
            }
        }, {}, {
            runtimeCacheInputs: ['echo runtime123', 'echo runtime456']
        });
        const hash = yield hasher.hash({
            target: { project: 'proj', target: 'build' },
            id: 'proj-build',
            overrides: { prop: 'prop-value' }
        });
        expect(hash).toContain('yarn.lock.hash'); //implicits
        expect(hash).toContain('file.hash'); //project files
        expect(hash).toContain('prop-value'); //overrides
        expect(hash).toContain('proj'); //project
        expect(hash).toContain('build'); //target
        expect(hash).toContain('runtime123'); //target
        expect(hash).toContain('runtime456'); //target
        done();
    }));
    it('should throw an error when failed to execute runtimeCacheInputs', () => __awaiter(void 0, void 0, void 0, function* () {
        const hasher = new hasher_1.Hasher({
            nodes: {},
            dependencies: {}
        }, {}, {
            runtimeCacheInputs: ['boom']
        });
        try {
            yield hasher.hash({
                target: { project: 'proj', target: 'build' },
                id: 'proj-build',
                overrides: {}
            });
            fail('Should not be here');
        }
        catch (e) {
            expect(e.message).toContain('Nx failed to execute runtimeCacheInputs defined in nx.json failed:');
            expect(e.message).toContain('boom:');
            expect(e.message).toContain(' not found');
        }
    }));
    it('should hash when circular dependencies', (done) => __awaiter(void 0, void 0, void 0, function* () {
        hashes['/filea'] = 'a.hash';
        hashes['/fileb'] = 'b.hash';
        const hasher = new hasher_1.Hasher({
            nodes: {
                proja: {
                    name: 'proja',
                    type: 'lib',
                    data: { files: [{ file: '/filea', ext: '.ts', mtime: 1 }] }
                },
                projb: {
                    name: 'projb',
                    type: 'lib',
                    data: { files: [{ file: '/fileb', ext: '.ts', mtime: 1 }] }
                }
            },
            dependencies: {
                proja: [{ source: 'proja', target: 'projb', type: 'static' }],
                projb: [{ source: 'projb', target: 'proja', type: 'static' }]
            }
        }, {}, {});
        const hasha = yield hasher.hash({
            target: { project: 'proja', target: 'build' },
            id: 'proja-build',
            overrides: { prop: 'prop-value' }
        });
        expect(hasha).toContain('yarn.lock.hash'); //implicits
        expect(hasha).toContain('a.hash'); //project files
        expect(hasha).toContain('b.hash'); //project files
        expect(hasha).toContain('prop-value'); //overrides
        expect(hasha).toContain('proj'); //project
        expect(hasha).toContain('build'); //target
        const hashb = yield hasher.hash({
            target: { project: 'projb', target: 'build' },
            id: 'projb-build',
            overrides: { prop: 'prop-value' }
        });
        expect(hashb).toContain('yarn.lock.hash'); //implicits
        expect(hashb).toContain('a.hash'); //project files
        expect(hashb).toContain('b.hash'); //project files
        expect(hashb).toContain('prop-value'); //overrides
        expect(hashb).toContain('proj'); //project
        expect(hashb).toContain('build'); //target
        done();
    }));
    it('should handle large binary files in a special way', (done) => __awaiter(void 0, void 0, void 0, function* () {
        fs.statSync.mockImplementation(f => {
            if (f === '/file')
                return { size: 1000000 * 5 + 1 };
            return { size: 100 };
        });
        hashes['/file'] = 'file.hash';
        const hasher = new hasher_1.Hasher({
            nodes: {
                proja: {
                    name: 'proj',
                    type: 'lib',
                    data: { files: [{ file: '/file', ext: '.ts', mtime: 1 }] }
                }
            },
            dependencies: {}
        }, {}, {});
        const hash = yield hasher.hash({
            target: { project: 'proja', target: 'build' },
            id: 'proja-build',
            overrides: { prop: 'prop-value' }
        });
        expect(hash).toContain('yarn.lock.hash'); //implicits
        expect(hash).toContain('5000001'); //project files
        expect(hash).toContain('prop-value'); //overrides
        expect(hash).toContain('proj'); //project
        expect(hash).toContain('build'); //target
        done();
    }));
});
